Bottom: 2aad87442f67375a1d96704ceb15db2efd6f265d
Top:    99d7f7cdedc2e0c10aa8852b0714b22ae9fe54c9
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-07-16 15:22:23 -0500

Add devpi service, a PyPi caching proxy


---
diff --git a/ansible-playbook/buildbot.yml b/ansible-playbook/buildbot.yml
index f5bf035e..b191985d 100644
--- a/ansible-playbook/buildbot.yml
+++ b/ansible-playbook/buildbot.yml
@@ -26,6 +26,23 @@
       new_user_email: buildbot@novatech-llc.com
       new_user_pass: "{{ lookup('password', '/dev/null') }}"
 
+- name: Setup Buildbot worker system
+  hosts: builder
+  become: true
+  become_user: root
+  become_method: sudo
+  tags:
+    - buildbot
+    - buildbot-worker
+  vars:
+    pypi_host: "{{ hostvars['buildbot-worker-devpi'].uri }}"
+    container_networks:
+      - name: "{{ worker_backend_network }}"
+        links:
+          - buildbot-worker-devpi:pypi
+  tasks:
+  - import_role: 
+      name: buildbot-worker-devpi
 - name: Gitlab public keys for buildbot containers
   hosts: daedalus
   become: true
diff --git a/roles/buildbot-worker-devpi/defaults/main.yml b/roles/buildbot-worker-devpi/defaults/main.yml
new file mode 100644
index 00000000..d443922a
--- /dev/null
+++ b/roles/buildbot-worker-devpi/defaults/main.yml
@@ -0,0 +1,30 @@
+---
+
+# environment passed to the container
+container_env:
+
+# the name of the container being started
+container_name: "{{ docker_name_prefix }}{{ hostname }}"
+
+# networks of the container
+container_networks:
+- name: "{{ container_name }}-network"
+
+# port configuration of the container
+container_port_args: []
+
+# volumes mounted within the container
+container_volumes:
+- "/devpi"
+
+# service hostname
+hostname: "{{ role_name }}"
+
+# build arguments of the image
+image_args:
+
+# directory of the image source
+image_dir: "{{ docker_projects_dir }}/{{ image_name }}"
+
+# name of the image being built
+image_name: "{{ docker_registry_username }}/{{ hostname }}"
diff --git a/roles/buildbot-worker-devpi/files/Dockerfile b/roles/buildbot-worker-devpi/files/Dockerfile
new file mode 100644
index 00000000..d66a40bb
--- /dev/null
+++ b/roles/buildbot-worker-devpi/files/Dockerfile
@@ -0,0 +1,50 @@
+FROM alpine:3.7 AS devpi-build
+
+ARG DEVPI_VERSION=4.6.0
+
+WORKDIR /wheels
+
+RUN apk add --no-cache \
+    alpine-sdk \
+    curl \
+    libffi-dev \
+    python-dev \
+    py-pip \
+&&  pip install --upgrade pip setuptools wheel \
+&&  pip install devpi_server==${DEVPI_VERSION} \
+&&  pip list --format freeze | grep -v '^devpi-server=' > wheels.txt \
+&&  pip wheel -r wheels.txt
+
+
+FROM alpine:3.7
+
+ARG DEVPI_VERSION=4.6.0
+
+COPY --from=devpi-build /wheels /wheels
+RUN apk add --no-cache \
+    dumb-init \
+    python \
+    py-pip \
+&&  pip install --upgrade pip setuptools \
+&&  pip install /wheels/*.whl \
+&&  pip install devpi-server==${DEVPI_VERSION} \
+&&  rm -r /root/.cache /wheels
+
+WORKDIR /devpi
+VOLUME /devpi
+
+ENV HOST=0.0.0.0
+ARG PORT=80
+ENV PORT=${PORT}
+EXPOSE ${PORT}
+
+ENTRYPOINT ["dumb-init"]
+
+CMD /usr/bin/devpi-server \
+    --serverdir /devpi \
+    --init; \
+    /usr/bin/devpi-server \
+    --serverdir /devpi \
+    --restrict-modify ''\
+    --host ${HOST} \
+    --port ${PORT}
diff --git a/roles/buildbot-worker-devpi/tasks/main.yml b/roles/buildbot-worker-devpi/tasks/main.yml
new file mode 100644
index 00000000..63e2e9ca
--- /dev/null
+++ b/roles/buildbot-worker-devpi/tasks/main.yml
@@ -0,0 +1,39 @@
+---
+- name: Create image build directory
+  file:
+    name: "{{ image_dir }}"
+    state: directory
+
+- name: Copy Dockerfile to server
+  copy:
+    src: Dockerfile
+    dest: "{{ image_dir }}"
+
+- name: Build devpi server image
+  docker_image:
+    name: "{{ image_name }}"
+    dockerfile: Dockerfile
+    path: "{{ image_dir }}"
+    force: true
+    state: present
+
+- name: Create internal devpi network
+  docker_network:
+    name: "{{ item.name }}"
+    state: present
+  loop: "{{container_networks}}"
+
+- name: Run devpi server
+  docker_container:
+    env: "{{ container_env }}"
+    hostname: "{{ hostname }}"
+    image: "{{ image_name }}"
+    name: "{{ container_name }}"
+    networks: "{{ container_networks }}"
+    ports: "{{ container_port_args }}"
+    state: started
+    volumes: "{{ container_volumes }}"
+
+- add_host:
+    name: "{{ hostname }}"
+    uri: "http://{{hostname}}/root/pypi/+simple/"
diff --git a/roles/buildbot-worker-devpi/vars/main.yml b/roles/buildbot-worker-devpi/vars/main.yml
new file mode 100644
index 00000000..e69de29b
